---
title: "Atlas integration"
author: "James Ashmore"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## Motivation

## Setup

Setup chunk options:

```{r}
knitr::opts_chunk$set(
  autodep    = TRUE,
  cache      = TRUE,
  cache.path = "cache/10-atlas-integration.Rmd/",
  dev        = "png",
  error      = FALSE,
  message    = FALSE,
  warning    = FALSE,
  fig.align = "center",
  out.width = '100%'
)
```

Load Bioconductor packages:

```{r}
library(batchelor)
library(EnsDb.Mmusculus.v79)
library(MouseGastrulationData)
library(org.Mm.eg.db)
library(pheatmap)
library(scater)
library(scran)
```

Load CRAN packages:

```{r}
library(msigdbr)
```

## Data

### Test data

Read experiment object:

```{r}
sce <- readRDS("output/09-cell-annotation.rds")
```

Clear metadata slot:

```{r}
metadata(sce) <- list()
```

Make rownames compatible:

```{r}
rownames(sce) <- rowData(sce)$gene_id
```

Clear rowData slot:

```{r}
rowData(sce) <- NULL
```

Replace rowRanges slot:

```{r}
rowRanges(sce) <- relist(GRanges(), PartitioningByEnd(integer(length(sce)), names = rownames(sce)))
```

Clear colnames slot:

```{r}
colnames(sce) <- NULL
```

Make colData compatible:

```{r}
colData(sce) <- DataFrame(
  cell = NA,
  barcode = NA,
  sample = 38, # reference data has 37 samples
  pool = NA,
  stage = NA,
  sequencing.batch = NA,
  theiler = NA,
  doub.density = NA,
  doublet = NA,
  cluster = NA,
  cluster.sub = NA,
  cluster.stage = NA,
  cluster.theiler = NA,
  stripped = NA,
  celltype = NA,
  colour = NA,
  sizeFactor = sizeFactors(sce),
  source = "test" # identify source of data
)
```

Clear reducedDims slot:

```{r}
reducedDims(sce) <- NULL
```

Clear altExp slot:

```{r}
altExp(sce) <- NULL
```

### Reference data

Load the mouse gastrulation data package:

```{r cache = FALSE}
library(MouseGastrulationData)
```

Subset atlas sample metadata by embryonic stage:

```{r cache = FALSE}
AtlasSampleMetadata <- subset(AtlasSampleMetadata, stage %in% c("E7.25", "E7.5", "E7.75", "E8.0", "E8.25"))
```

Retrieve processed samples from embryo atlas data:

```{r cache = FALSE}
ref <- EmbryoAtlasData(type = "processed", samples = AtlasSampleMetadata$sample)
```

Append source label to colData:

```{r cache = FALSE}
ref$source <- "reference"
```

Compute log-normalized expression values:

```{r cache = FALSE}
ref <- logNormCounts(ref)
```

Remove cells without a celltype label:

```{r cache = FALSE}
use <- !is.na(ref$celltype)

ref <- ref[, use]
```

Remove cells marked as stripped nuclei:

```{r cache = FALSE}
use <- !ref$stripped

ref <- ref[, use]
```

Remove cells marked as doublets:

```{r cache = FALSE}
use <- !ref$doublet

ref <- ref[, use]
```

Clear metadata slot:

```{r cache = FALSE}
metadata(ref) <- list()
```

Make rownames compatible:

```{r cache = FALSE}
rownames(ref) <- rowData(ref)$ENSEMBL
```

Clear rowData slot:

```{r cache = FALSE}
rowData(ref) <- NULL
```

Replace rowRanges slot:

```{r cache = FALSE}
rowRanges(ref) <- relist(GRanges(), PartitioningByEnd(integer(length(ref)), names = rownames(ref)))
```

Clear colnames slot:

```{r cache = FALSE}
colnames(ref) <- NULL
```

Clear reducedDims slot:

```{r cache = FALSE}
reducedDims(ref) <- NULL
```

Clear altExp slot:

```{r cache = FALSE}
altExp(ref) <- NULL
```

Split a SingleCellExperiment based on the column data:

```{r cache = FALSE}
splitColData <- function(x, f) {
  
  i <- split(seq_along(f), f)

  v <- vector(mode = "list", length = length(i))
  
  names(v) <- names(i)
  
  for (n in names(i)) { v[[n]] <- x[, i[[n]]] }
  
  return(v)

}
```

Split experiment data by the sample column:

```{r cache = FALSE}
ref <- splitColData(ref, ref$sample)
```

## Integration

### Data processing

Create list of experiment data:

```{r cache = FALSE}
all <- c(ref, "38" = sce)
```

Delete large objects and trigger garbage collection:

```{r}
rm(sce, ref); gc()
```

Subset data to common features:

```{r cache = FALSE}
ids <- Reduce(intersect, lapply(all, rownames))

all <- lapply(all, "[", i = ids)
```

### Variance modelling

Model the per-gene variance:

```{r cache = FALSE}
dec <- lapply(all, modelGeneVar)
```

Combine variance decompositions:

```{r}
dec <- do.call(combineVar, dec)

rownames(dec) <- Reduce(intersect, lapply(all, rownames))
```

### Feature selection

Create a blacklist of genes to ignore for feature selection:

```{r}
dec$blacklist <- FALSE
```

Append Xist gene to blacklist:

```{r}
idx <- which(rownames(dec) == "ENSMUSG00000086503")

dec$blacklist[idx] <- TRUE
```

Append Y chromosome genes to blacklist:

```{r}
key <- keys(EnsDb.Mmusculus.v79, keytype = "GENEID", filter = SeqNameFilter("Y"))

idx <- which(rownames(dec) %in% key)

dec$blacklist[idx] <- TRUE
```

Append MT chromosome genes to blacklist:

```{r}
key <- keys(EnsDb.Mmusculus.v79, keytype = "GENEID", filter = SeqNameFilter("MT"))

idx <- which(rownames(dec) %in% key)

dec$blacklist[idx] <- TRUE
```

Append cell-cycle genes to blacklist:

```{r}
sel <- select(org.Mm.eg.db, keys = "GO:0007049", keytype = "GOALL", column = "ENSEMBL")

idx <- which(rownames(dec) %in% sel$ENSEMBL)

dec$blacklist[idx] <- TRUE
```

Append ribosomal genes to blacklist:

```{r}
msd <- msigdbr(species = "Mus musculus", category = "C2")

msd <- subset(msd, gs_name == "KEGG_RIBOSOME")

sel <- select(org.Mm.eg.db, keys = msd$gene_symbol, keytype = "SYMBOL", column = "ENSEMBL")

idx <- which(rownames(dec) %in% sel$ENSEMBL)

dec$blacklist[idx] <- TRUE
```

Identify highly variable genes:

```{r}
hvg <- subset(dec, blacklist == FALSE)

hvg <- getTopHVGs(hvg, var.field = "bio", var.threshold = 0, fdr.field = "p.value", fdr.threshold = 0.05)
```

### Batch normalization

Perform scaling normalization:

```{r cache = FALSE}
all <- do.call(multiBatchNorm, all)
```

Combine objects from experiments:

```{r cache = FALSE}
all <- do.call(cbind, all)
```

### Batch integration

Create sequential merge order:

```{r}
dat <- subset(AtlasSampleMetadata, stage %in% all$stage)

lvl <- c("E6.5", "E6.75", "E7.0", "mixed_gastrulation", "E7.25", "E7.5", "E7.75", "E8.0", "E8.25", "E8.5")

dat$stage <- factor(dat$stage, levels = intersect(lvl, all$stage))

dat <- dat[order(dat$stage, dat$ncells, decreasing = TRUE), ]
  
idx <- c(as.character(dat$sample), "38")
```

Correct experiments using nearest neighbours:

```{r cache = FALSE}
set.seed(304391910)

mnn <- correctExperiments(
  all,
  batch = all$sample,
  subset.row = hvg,
  correct.all = TRUE,
  PARAM = FastMnnParam(merge.order = idx)
)
```

Delete large objects and trigger garbage collection:

```{r}
rm(all); gc()
```

Replace rowData slot:

```{r cache = FALSE}
rowData(mnn) <- DataFrame(
  gene_id = rownames(mnn),
  gene_name = mapIds(EnsDb.Mmusculus.v79, keys = rownames(mnn), column = "GENENAME", keytype = "GENEID"),
  rotation = I(rowData(mnn)$rotation)
)
```

Store combined variance model:

```{r cache = FALSE}
rownames(dec) <- rownames(mnn)

metadata(mnn)$combineVar <- dec
```

Store highly variable genes:

```{r cache = FALSE}
metadata(mnn)$getTopHVGs <- hvg
```

## Diagnostics

Check proportion of lost variance:

```{r cache = FALSE}
var <- metadata(mnn)$merge.info$lost.var

col <- colorRampPalette(RColorBrewer::brewer.pal(n = 5, name = "Reds"))(100)

brk <- seq(0, max(var), length.out = 101)

pheatmap(var[, idx], color = col, display_numbers = TRUE, number_format = "%.2f", cluster_cols = FALSE, cluster_rows = FALSE, angle_col = 0)
```

Cluster on the corrected PCA to obtain a partitioning of the cells:

```{r cache = FALSE}
snn <- buildSNNGraph(mnn, type = "jaccard", use.dimred = "corrected")

com <- igraph::cluster_louvain(snn)
```

Tabulate number of cells from multiple batches within each cluster:

```{r cache = FALSE}
tab <- table(Cluster = com$membership, Batch = mnn$batch)

col <- colorRampPalette(RColorBrewer::brewer.pal(n = 5, name = "Blues"))(100)

brk <- seq(0, max(tab), length.out = 101)

pheatmap(tab[, idx], color = col, display_numbers = TRUE, number_format = "%d", cluster_cols = FALSE, cluster_rows = FALSE, angle_col = 0)
```

Perform TSNE on the corrected PCA matrix:

```{r cache = FALSE}
set.seed(42)

mnn <- runTSNE(mnn, dimred = "corrected", perplexity = 120) # same settings as reference analysis
```

Visualize the corrected PCA using a TSNE plot:

```{r cache = FALSE, fig.width = 9, fig.height = 9}
ggcells(mnn, aes(TSNE.1, TSNE.2, colour = celltype)) + 
  geom_point(size = 0.1, show.legend = FALSE) + 
  scale_colour_manual(values = EmbryoCelltypeColours, na.value = "#000000") + 
  labs(x = "UMAP 1", y = "UMAP 2") + 
  facet_wrap(~ sample) + 
  theme_bw() + 
  theme(aspect.ratio = 1)
```

Perform UMAP on the corrected PCA matrix:

```{r cache = FALSE}
set.seed(42)

mnn <- runUMAP(mnn, dimred = "corrected", n_neighbors = 20, min_dist = 0.7) # same settings as reference analysis
```

Visualize the corrected PCA using a UMAP plot:

```{r cache = FALSE, fig.width = 9, fig.height = 9}
ggcells(mnn, aes(UMAP.1, UMAP.2, colour = celltype)) + 
  geom_point(size = 0.1, show.legend = FALSE) + 
  scale_colour_manual(values = EmbryoCelltypeColours, na.value = "#000000") + 
  labs(x = "UMAP 1", y = "UMAP 2") + 
  facet_wrap(~ sample) + 
  theme_bw() + 
  theme(aspect.ratio = 1)
```

## Summary

### Output

Save integrated experiment object to disk:

```{r cache = FALSE}
saveRDS(mnn, file = "output/10-atlas-integration.rds")
```

### Session

Print session information:

```{r}
sessionInfo()
```
